Task views

cyril-steimer opened the pull request:
A few changes were made here:
- Can now store and load Json from file (I needed to create a views folder in the DebugBuild folder, is there a way to do this without extra code, and without others having to manually create it?)
- The names to create a new ViewItem must now be unique.
- In the menu, when typing a name referring to a stored ViewItem, loads that.
- Added a command to add nodes by its name. I am again using the AllTreeManagers first TreeManager to access its root - is there a way to find the "current root"? It also uses AutoComplete in a way similar to the creating methods/classes thing, there is no need to spell the entire name. Separating parts of the name is done by _, using dots didn't work somehow.

cyril-steimer committed:
Adding extra info method for usages.

cyril-steimer committed:
Merge remote-tracking branch 'upstream/development' into task-views

cyril-steimer committed:
Notify ViewItem about removal of items/nodes.

cyril-steimer committed:
Add Json loading/storing to ViewItemManager

cyril-steimer committed:
Name must be unique, add function to load/create

cyril-steimer committed:
Add new command to add to view by name

cyril-steimer committed:
Merge remote-tracking branch 'upstream/development' into task-views

cyril-steimer commented:
Ah, and another question. What do you think is a good way on where to add new items. A quite trivial solution would be to allow the user to specify the position and the rendering purpose of the new node in the add command (and of course use default arguments if he doesn't).

Another solution would be to use the position closest to the cursor when the command was opened. The question is, is that possible? Meaning is there an easy way to 
a) find the position where the command prompt was created and 
b) find which item in the grid is under that position?

cyril-steimer committed:
Add command to save a view.

cyril-steimer committed:
Add command to project

dimitar-asenov commented:
First, let me answer your questions:
- Please make sure to programatically create the `views` directory if it doesn't exist. Obviously this will require some extra code, but it's a line or two: http://stackoverflow.com/questions/2241808/checking-if-a-folder-exists-and-creating-folders-in-qt-c
- In a command you have a `target` item. You can get the "appropriate" manager by calling: `target->findAncestorWithNode()->manager()`. Keep in mind, that in theory both of the calls may return `nullptr`.
- Regarding adding new items, this should be as convenient for the user as possible. So manually specifying the position seems totally like the wrong thing to do. Perhaps we can have that command option, but just for completeness. Using the cursor is possible and is the way to go. In particular you will notice that it is possible to navigate (using the arrows keys or the mouse) to the borders of the grid of the view. Once on a border you can hit escape and execute a command. The command has a `cursor` argument which you can use to determine precisely which position is selected. Hitting escape to execute a command might be OK, but we can make the process even nicer. E.g. in the interaction handler for the ViewItem, we could make it so that if a border is selected, pressing enter (easier than escape) automatically brings up the command menu, and not only that but also automatically fills in the command that is needed to open nodes. So that the only thing that the user would have to do is to press Enter and then type the name of the node they are looking for.

dimitar-asenov commented:
Hmm, regarding the second point, if you don't need the manager then just call `root()`:

```
target->findAncestorWithNode()->root();
```

dimitar-asenov commented on InteractionBase/src/handlers/GenericHandler.cpp:
```diff
@@ -424,6 +425,8 @@ void GenericHandler::keyPressEvent(Visualization::Item *target, QKeyEvent *event
 			// Check if the parent of the node is a list and if so, delete this node
 			if (auto list = DCast<Model::List>(node->parent()))
 			{
+				for (auto view : target->scene()->viewItems()->viewItemsAsList())
```
This is definitely the wrong place for doing this. There are many ways nodes can be deleted and we need a more generic mechanism for dealing with this. What's wrong with using `Item::removeFromScene`?

cyril-steimer replied:
I also put it in Item::removeFromScene.

So the reason I put something here as well is as follows: If a node is deleted, then the corresponding InfoNodes should be deleted. However, in the Viewitem, the same node might be visualized by many different items. Therefore, doing this when removing the item itself might be a too aggressive way to remove them (the InfoNode could refer to another visualization of the same node). Any ideas how to solve this?

I put code in Item::removeFromScene as well.

dimitar-asenov replied:
There is `Scene::nodesUpdated` which gets called whenever some nodes have been updated. You should try to work with that method. It receives a list of updated nodes. I'm not sure if this list will contain deleted nodes, or only the parents of deleted nodes, so you'd have to check. In any case, don't do too much work directly in this method. Just call something inside `ViewItemManager` that will do most of the work.

dimitar-asenov commented on InteractionBase/src/vis/ViewSwitcherMenu.cpp:
```diff
@@ -87,7 +87,7 @@ void ViewSwitcherMenu::endFocusMode(Visualization::Item *target)
 		auto nameAfter = entry->nameField()->text();
 		if (scene()->viewItems()->viewItem(nameAfter) == nullptr && nameAfter != nameBefore_)
-			scene()->viewItems()->newViewItem(nameAfter, pos);
+			scene()->viewItems()->createIfNotExists(nameAfter, pos);
```
I don't like this name, I guess mainly because of the `Not`. What about:
- `createIfNew()` // not so great to have an `If` in a method name
- `createOrOpen()` // I guess you don't really open anything, but just a thought
- something else that is more straight forward and precise

dimitar-asenov commented on OOInteraction/src/commands/CAddNodeToViewByName.cpp:
```diff
@@ -0,0 +1,155 @@
+		return new Interaction::CommandResult(new Interaction::CommandError("Please specify a node to add"));
+
+	if (auto node = findNode(commandTokens[1],
+			Model::AllTreeManagers::instance().loadedManagers().first()->root()))
```
as mentioned in the front page use `target` to find the proper root

cyril-steimer replied:
I decided to instead just loop through all the managers - this should also allow adding nodes from other projects. One question: Is the identifier of a node unique within all projects or only a single project?

dimitar-asenov replied:
well, it should be unique within all projects, but I wouldn't necessarily make my code dependent on this fact. Why is that important?

cyril-steimer replied:
Essentially, if it is unique, then in the JSON we need to only store the id of the node and can have nodes from multiple projects in a view. On the other hand, if it isn't, we need to also store the project's name itself.

dimitar-asenov replied:
Alright, just keep the ID for now.

dimitar-asenov commented on OOInteraction/src/commands/CAddNodeToViewByName.cpp:
```diff
@@ -0,0 +1,155 @@
+	if (auto node = findNode(commandTokens[1],
+			Model::AllTreeManagers::instance().loadedManagers().first()->root()))
+	{
+		source->scene()->currentViewItem()->insertNode(node);
```
always use `target` instead of `source`, unless you have an explicit need for using `source` (which you don't here)

dimitar-asenov commented on OOInteraction/src/commands/CAddNodeToViewByName.cpp:
```diff
@@ -0,0 +1,155 @@
+	{
+		auto nodeName = textSoFar.trimmed().mid(4);
+		auto names = findNames(nodeName.split("_"),
+						Model::AllTreeManagers::instance().loadedManagers().first()->root());
```
also here you can change to the current root.

On the other hand, we should think of a way to handle multiple trees. By default we should search for nodes in the current project, but we might also want to open a node from a different project. We must support that somehow. perhaps via an extra optional argument that indicates the project name.

dimitar-asenov commented on OOInteraction/src/commands/CAddNodeToViewByName.cpp:
```diff
@@ -0,0 +1,155 @@
+	if (textSoFar.trimmed().startsWith("add ", Qt::CaseInsensitive))
+	{
+		auto nodeName = textSoFar.trimmed().mid(4);
+		auto names = findNames(nodeName.split("_"),
```
we should use . instead of _. I think the reason why it might not have worked is that . (unlike _ ) results in a separate token: so `a_b` is one token whereas `a.b` are three tokens. Could you try this and see if you can adapt your code?

dimitar-asenov commented on OOInteraction/src/commands/CAddNodeToViewByName.cpp:
```diff
@@ -0,0 +1,155 @@
+	auto declaration = DCast<OOModel::Declaration>(root);
+	QStringList result;
+	//If it isn't a declaration, just pass it on
+	if (!declaration)
```
This is not the right way to search for names, mainly because it is specific to OO. You should not care at all about declarations or other things. What you should do is just traverse the tree and for each node use the `node->definesSymbol` and `node->symbolName`.

dimitar-asenov replied:
Now that I see this, I'm wondering whether this command should be in `OOInteraction` at all. We want this to work in general. Do we need to know about OO concepts at all here? Can't we just use the normal symbols and put this command in `InteractionBase` ?

cyril-steimer replied:
I moved it to InteractionBase. Furthermore, I also moved the existing AddNodeToView command which requires to select the node itself before adding, and also moved it there, using the symbolName stuff as well.

dimitar-asenov replied:
Good.

dimitar-asenov commented on OOInteraction/src/commands/CAddNodeToViewByName.cpp:
```diff
@@ -0,0 +1,155 @@
+
+OOModel::Declaration* CAddNodeToViewByName::findNode(QString fullyQualifiedName, Model::Node* root)
+{
+	//TODO can't use "." for some reason (command unknown..)
```
If you try the suggestion I had with the different tokens, and it still doesn't work, let me know.

dimitar-asenov commented on OOInteraction/src/commands/CAddNodeToViewByName.cpp:
```diff
@@ -0,0 +1,155 @@
+	return nullptr;
+}
+
+bool CAddNodeToViewByName::isSuggestable(OOModel::Declaration *declaration)
```
To achieve something similar, using just generic concepts (without relying on OO specific) use the `node->symbolType` and check for method and container.

dimitar-asenov commented on OOVisualization/src/info/InfoMethods.cpp:
```diff
@@ -37,12 +36,35 @@ QString InfoMethods::numberOfCallees(Model::Node *node)
+	else if (auto clazz = DCast<OOModel::Class>(node))
+	{
+		QSet<Model::Node*> result;
+		auto top = clazz->root();
```
don't use hacky names such as `clazz`. Say `someClass`

dimitar-asenov commented on OOVisualization/src/info/InfoMethods.cpp:
```diff
@@ -37,12 +36,35 @@ QString InfoMethods::numberOfCallees(Model::Node *node)
+		{
+			auto check = toCheck.takeLast();
+			if (auto expr = DCast<OOModel::ReferenceExpression>(check))
+				if (OOModel::Class::expressionToClass(expr) == clazz)
```
Better check: `expr->target() == someClass`

dimitar-asenov commented on OOInteraction/src/commands/CAddNodeToViewByName.cpp:
```diff
@@ -0,0 +1,155 @@
+	if (!declaration)
+		for (auto child : root->children())
+			result.append(findNames(nameParts, child));
+	//If it is a declaration, and the name matches, take the part of name it matches and search for the rest
```
When you write comments/methods like these, please use more white space. In this method please leave an empty line before all the comments.

dimitar-asenov commented on VisualizationBase/src/ViewItemManager.cpp:
```diff
@@ -67,6 +68,23 @@ ViewItem* ViewItemManager::viewItem(const QString name)
 	return nullptr;
 }
 
+void ViewItemManager::addViewItem(ViewItem *view, QPoint position)
```
Hmm, I see that you moved this method earlier in the file. What's the reason for this?

dimitar-asenov commented on VisualizationBase/src/ViewItemManager.cpp:
```diff
@@ -121,6 +124,57 @@ void ViewItemManager::removeAllViewItems()
 	currentViewItem_ = nullptr;
 }
 
+ViewItem* ViewItemManager::createIfNotExists(const QString name, QPoint position)
```
Hmm, so this function does return the existing view item if any? Perhaps a better name for it would be `createOrGet`. Although, the `position` parameter isn't used in case the item already exists.

So thinking about this, I don't quite like it:
- We already have a `newViewItem()` method and now we introduce a second one that essentially also creates a new view item. Both of them are `public` which essentially obscures the interface quite a bit.
- We should have only one method that creates a new view. This method should be public. Behind the scenes it could try to load the view or make a new one or whatever.
- I'm not sure if we should combine the creation method with a method that just returns an existing view. Why don't we require that creating a new view with a specific name/position should have a unique name/position and make the client of this method check that? What exactly are the use cases of this method?

cyril-steimer replied:
Alright, I changed the interface a bit. I removed that new method, and made addViewItem() private.  Furthermore, I added a position argument the loadView, such that loadView now doesn't only return the ViewItem, but also adds it to the items internally.

dimitar-asenov commented on VisualizationBase/src/ViewItemManager.cpp:
```diff
@@ -121,6 +124,57 @@ void ViewItemManager::removeAllViewItems()
+{
+	//Try seeing if it is opened already
+	if (auto view = viewItem(name))
+		return view;
```
Here and everywhere else, prefer putting very short `if`/`for`/`while` bodies on the same line as the statement header.

dimitar-asenov commented on VisualizationBase/src/ViewItemManager.cpp:
```diff
@@ -121,6 +124,57 @@ void ViewItemManager::removeAllViewItems()
+	file.open(QIODevice::WriteOnly);
+	QTextStream write(&file);
+	write << json;
+	file.close();
```
You don't need to do this. The file will be automatically closed at the end of this method.

dimitar-asenov commented on VisualizationBase/src/ViewItemManager.cpp:
```diff
@@ -121,6 +124,57 @@ void ViewItemManager::removeAllViewItems()
+	return view;
+}
+
+QList<ViewItem*> ViewItemManager::viewItemsAsList() const
```
I don't think you should have this method. I will comment more in the place you use it below.

dimitar-asenov commented on VisualizationBase/src/ViewItemManager.h:
```diff
@@ -44,19 +48,27 @@ class VISUALIZATIONBASE_API ViewItemManager
-		void addViewItem(ViewItem* view, QPoint position = QPoint(-1, -1));
 		ViewItem* newViewItem(const QString name = QString(), QPoint position = QPoint(-1, -1));
 		ViewItem* viewItem(const QString name);
+		void addViewItem(ViewItem* view, QPoint position = QPoint(-1, -1));
```
This class has way too many public methods and its API could be improved:
- It seems like we have not one but three ways to create views:
  - `newViewItem`
  - `addViewItem`
  - `createIfNotExist`
  - We should have exactly one public method to do this
- `viewItemsAsList` seems unnecessary to me
- We should reconsider having public `saveView`/`loadView` methods.
  - `loadView` is only called by the manager itself. Why not make it private and make the API leaner?
- `saveView` is only called with the current view as an argument. I think it's better if we make the current implementation private and have a new public method `saveCurrentView` or perhaps even `saveAllViews`. It might even be a good idea to save these automatically every time something changes on the view.

dimitar-asenov commented on VisualizationBase/src/items/Item.cpp:
```diff
@@ -419,6 +421,10 @@ void Item::removeFromScene()
 		scene()->setItemIsSensitiveToScale(this, false);
 
+		//Notify all views about deletion of the item
+		for (auto view : scene()->viewItems()->viewItemsAsList())
```
Instead of manually iterating over all the views, just call a `notify` method on the `viewManager`. This way you let the manager decide what to do, which is the correct place to the iteration. This will also remove the need to include the `ViewItem.h` file above.

dimitar-asenov commented on VisualizationBase/src/items/ViewItem.cpp:
```diff
@@ -142,6 +130,63 @@ Model::Node* ViewItem::nodeAt(int column, int row) const
 	return nodes_[column][row];
 }
 
+void ViewItem::notifyAboutRemoval(Item *item)
```
I'm not happy about the names of this and the next method. `notify` typically is used in contexts where the signals and slots mechanism is used, which is not the case here. Could you please rename them to:
- `cleanupRemovedNode`
- `cleanupRemovedItem`

or some other more appropriate name.

dimitar-asenov commented on VisualizationBase/src/items/ViewItem.cpp:
```diff
@@ -142,6 +130,63 @@ Model::Node* ViewItem::nodeAt(int column, int row) const
+	//If somebody's spacing depends on the node or a child, remove
+	//the node's spacing target
+	QList<ViewItemNode*> spacingNodes;
+	for (auto node : allNodes())
```
Ouch ouch, a local variable with the same name as the argument. This is asking for trouble. Please, rename this.

dimitar-asenov commented on VisualizationBase/src/items/ViewItem.cpp:
```diff
@@ -142,6 +130,63 @@ Model::Node* ViewItem::nodeAt(int column, int row) const
+	QList<Model::Node*> toCheck{node};
+	while (!toCheck.isEmpty())
+	{
+		auto current = toCheck.takeFirst();
```
use `takeLast` to get a depth-first traversal and reduce memory consumption.

dimitar-asenov commented on VisualizationBase/src/items/ViewItem.cpp:
```diff
@@ -142,6 +130,63 @@ Model::Node* ViewItem::nodeAt(int column, int row) const
+			}
+	}
+	//Remove all the info nodes depending on the node
+	QList<ViewItemNode*> infoNodeParents;
```
This code is almost identical to the for loop above that gets you all the spacing nodes. Put this stuff in a separate templated method `allNodesOfType`.

dimitar-asenov commented on VisualizationBase/src/items/ViewItem.cpp:
```diff
@@ -142,6 +130,63 @@ Model::Node* ViewItem::nodeAt(int column, int row) const
+	toCheck.append(node);
+	while (!toCheck.isEmpty())
+	{
+		auto current = toCheck.takeFirst();
```
takeLast

dimitar-asenov commented on VisualizationBase/src/items/ViewItem.cpp:
```diff
@@ -142,6 +130,63 @@ Model::Node* ViewItem::nodeAt(int column, int row) const
+	while (!toCheck.isEmpty())
+	{
+		auto current = toCheck.takeFirst();
+		toCheck.append(current->children());
```
Hmm, why not just reuse the while loop above and handle both spacing and info nodes at once? This way you save yourself code duplication and redundant iteration.

dimitar-asenov commented on VisualizationBase/src/items/ViewItem.h:
```diff
@@ -55,20 +55,68 @@ class VISUALIZATIONBASE_API ViewItem : public Super<DeclarativeItem<ViewItem>> {
+		/**
+		 * Returns the full layer name of the local arrow layer, as used in the overlay which
+		 * exists in the scene itself.
+		 */
```
Thanks for putting all these comments here.

cyril-steimer committed:
Rename method createIfNotExists

cyril-steimer committed:
Rename local variable clazz

cyril-steimer committed:
Fix command, move to InteractionBase

cyril-steimer committed:
Move command to InteractionBase

cyril-steimer committed:
Changes to ViewItemManager organization

cyril-steimer committed:
Small changes

cyril-steimer committed:
Bugfix to restore arrows from Json correctly

cyril-steimer committed:
Create folder if it doesn't exist.

cyril-steimer committed:
Correctly handles hidden arrow layers

cyril-steimer committed:
Improve the toggle arrow layer command

cyril-steimer committed:
Improve the toggle info entry command

cyril-steimer committed:
Can add nodes of different projects.

dimitar-asenov commented on InteractionBase/src/commands/CSaveView.cpp:
```diff
@@ -0,0 +1,63 @@
+	//Try to use an existing manager if possible
+	if (source->findAncestorWithNode() && source->findAncestorWithNode()->node()->manager())
+	{
+		source->scene()->viewItems()->saveView(source->scene()->currentViewItem(),
```
here and in all the other places in your code, could you please make sure to use `target` and not `source`?

cyril-steimer replied:
I had to check again, but there is actually a reason I'm using source and not target here: Target is always the SceneHandlerItem, which has no parent with a node (and therefore I can't find the manager with it). Source however is the actual item which is part of the "main item tree".

cyril-steimer replied:
Should I just register these commands which aren't default commands with a different handler and then use target? Or leave it as is?

dimitar-asenov replied:
Ok so this is a legitimate use of `source`, sorry for the confusion. In general there are two options:
- Register the commands with the specific types of Items that they can handle. In that case use `target`. Use this approach for commands which are very specific to certain items (typically they can work on a single item only).
- Register the commands on a more top-level item (i.e. SceneHandlerItem) and use `source` if necessary. Use this approach for more generic commands which might need to work on a wide variety of nodes.

In this particular case, we want the save command to work on pretty much anything, so your existing solution is the preferred one.

Just make sure to look over the rest of the commands and use `source`/`target` as appropriate.

dimitar-asenov commented on InteractionBase/src/vis/ViewSwitcherMenu.cpp:
```diff
@@ -87,7 +87,10 @@ void ViewSwitcherMenu::endFocusMode(Visualization::Item *target)
 		if (scene()->viewItems()->viewItem(nameAfter) == nullptr && nameAfter != nameBefore_)
-			scene()->viewItems()->newViewItem(nameAfter, pos);
+		{
+			if (scene()->viewItems()->loadView(nameAfter, pos)) {}
```
Hmm I thought I commented on this one, but I don't see my comment. Use `if (! ...) ... //without else`

dimitar-asenov commented on VisualizationBase/src/ViewItemManager.h:
```diff
@@ -52,11 +56,22 @@ class VISUALIZATIONBASE_API ViewItemManager
 		QVector<QVector<ViewItem*>> viewItems() const;
+		QList<ViewItem*> viewItemsAsList() const;
+
+		void cleanupRemovedItem(Visualization::Item* removedItem);
```
This is not exactly a public method. Better make it private and declare whoever is calling it as a friend. friend declarations should appear on top of the private section in a class.

dimitar-asenov commented on VisualizationBase/src/items/ViewItem.cpp:
```diff
@@ -169,22 +206,43 @@ QList<QPair<Item*, Item*>> ViewItem::arrowsForLayer(QString layer)
+			bool addedArrow = false;
+			for (auto from : allFrom)
+				for (auto to : allTo)
+					//TODO@cyril The node is rendered as a RootItem -> needs a hack here
```
Tell me more about this. What's the cause of this issue?

cyril-steimer replied:
Well, essentially what happens is that in the renderer we render ViewItems as RootItems if I remember correctly. So what happens here, is that there would be two items visualizing the same ViewItemNode (the RootItem and the VViewItemNode), and arrows would exist multiple times. That causes an issue when an item is then deleted.

dimitar-asenov commented on VisualizationBase/src/items/ViewItem.cpp:
```diff
@@ -204,6 +262,11 @@ void ViewItem::determineChildren()
 				}
 		}
 	}
+	//Must make sure to add the necessary arrow layers here, as sometimes they are added
```
Sometimes? but not always? What's the reason? We should find the root causes of such behaviors and not just implement workarounds that seem to do the trick.

dimitar-asenov replied:
Could you, please comment on this.

cyril-steimer replied:
Ah sorry, I thought I commented on this, but I think it was when my internet was broken.
Anyway, there are two cases when we add arrows:
1) When working normally with a task view -> then the view is always already added to the scene
2) When loading a task view from memory -> then it is initialized from Json before being added to the scene

dimitar-asenov commented on VisualizationBase/src/overlays/OverlayGroup.h:
```diff
@@ -44,6 +44,7 @@ class VISUALIZATIONBASE_API OverlayGroup {
 		void hide();
 		void show();
 		void toggle();
+		bool isVisible() const;
```
Thanks for this.

dimitar-asenov commented:
Ok so there are still a few things but after that I should be able to merge:
- [x] `source` - > `target` everywhere
- [x] use `Scene::nodesUpdated` and remove `viewItemsAsList()`
- [x] use `if (!...) ...//without else`
- [x] make `cleanupRemoveItem` private
- [x] I posted a few small questions regarding bugs and workarounds, depending on what you find we might need to change the code a bit.

cyril-steimer committed:
Fix how items/nodes are removed and handled by ViewItem

cyril-steimer commented:
So I think apart from the source -> target (for which I wrote a separate comment), I fixed these issues. I didn't start yet choosing more intelligently where to put the new item, but I think that's not necessary yet to merge it. Other than that, I believe I changed everything you mentioned or at least commented on why I didn't.

cyril-steimer committed:
Prepare ViewItem for improved deletion handling

cyril-steimer committed:
Improvement to removing arrows.

dimitar-asenov commented:
Alright, I committed the change to make `Scene::nodesUpdated` also receive what nodes have been deleted (including child nodes). So please make use of that now, and hopefully that will work well.

cyril-steimer committed:
Merge remote-tracking branch 'upstream/development' into task-views

cyril-steimer committed:
Use new removedNodes from Scene for better deletion handling

cyril-steimer committed:
Use default spacing height if no spacing target

cyril-steimer commented:
Alright, great. I used it and made the necessary changes, seems like everything works (particularly, so do the scenarios we discussed on Skype).

dimitar-asenov commented:
Great, thanks for putting in all this effort and making sure that things are made properly :)
